File: iOSDemoApp/SplashViewController.m
//
//  SplashViewController.m
//  SeaCatClient
//
//  Created by Ales Teska on 28/02/16.
//
//

#import "SplashViewController.h"
#import "SeaCatClient/SeaCatClient.h"

@interface SplashViewController ()

@end

@implementation SplashViewController
{
    NSTimer * periodicTimer;
}

///

- (void)viewWillAppear:(BOOL)animated
{
    if ([SeaCatClient isReady]) [self dismissViewControllerAnimated:YES completion:nil];
    [super viewWillAppear:animated];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    [SeaCatClient addObserver:self selector:@selector(onStateChanged) name:SeaCat_Notification_StateChanged];
    periodicTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(onStateChanged) userInfo:nil repeats:YES];
    [self onStateChanged];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [SeaCatClient removeObserver:self];

    [periodicTimer invalidate];
    periodicTimer = nil;

    [super viewWillDisappear:animated];
}

///

- (void)onStateChanged
{
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        NSString * state = [SeaCatClient getState];
        [_stateItem setTitle:state];
        
        _clientTagLabel.text = [SeaCatClient getClientTag];
        
        // SeaCat is ready
        if ([SeaCatClient isReady]) [self dismissViewControllerAnimated:YES completion:nil];
    }];
}

///

- (IBAction)onResetAction:(id)sender
{
    UIAlertController *myAlertController = [UIAlertController
        alertControllerWithTitle:@"Reset identity"
        message: @"Are you sure?"
        preferredStyle:UIAlertControllerStyleAlert
    ];
    
    UIAlertAction* act_yes = [UIAlertAction
        actionWithTitle:@"Yes"
        style:UIAlertActionStyleDestructive
        handler:^(UIAlertAction * action)
        {
            [SeaCatClient reset];
            [myAlertController dismissViewControllerAnimated:YES completion:nil];
        }
    ];
    [myAlertController addAction:act_yes];
    
    UIAlertAction* act_no = [UIAlertAction
        actionWithTitle:@"No"
        style:UIAlertActionStyleCancel
        handler:^(UIAlertAction * action)
        {
            [myAlertController dismissViewControllerAnimated:YES completion:nil];
        }
    ];
    [myAlertController addAction:act_no];
    
    [self presentViewController:myAlertController animated:YES completion:nil];
}

@end

File: iOSwiftDemoApp/SplashViewController.swift
//
//  SplashViewController.swift
//  iOSwiftDemoApp
//
//  Created by Ales Teska on 24.4.18.
//

import UIKit
import SeaCatClient

class SplashViewController: UIViewController {
    
    @IBOutlet weak var clientTagLabel: UILabel!
    @IBOutlet weak var stateLabel: UIBarButtonItem!
    var periodicTimer: Timer!
    
    override func viewWillAppear(_ animated: Bool) {
        if (SeaCatClient.isReady())
        {
            self.dismiss(animated: true, completion: nil);
        }

        return super.viewWillAppear(animated);
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated);
        
        SeaCatClient.addObserver(self, selector:#selector(self.onStateChanged), name:SeaCat_Notification_StateChanged);
        self.periodicTimer = Timer.scheduledTimer(timeInterval:1.0, target:self, selector:#selector(self.onStateChanged), userInfo:nil, repeats:true);
        self.onStateChanged();
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        SeaCatClient.removeObserver(self);
        
        self.periodicTimer.invalidate();
        self.periodicTimer = nil;
        
        super.viewWillDisappear(animated);
    }
    
    @objc func onStateChanged()
    {
        OperationQueue.main.addOperation {
            self.stateLabel.title = SeaCatClient.getState();
            self.clientTagLabel.text = SeaCatClient.getClientTag();

            if (SeaCatClient.isReady())
            {
                self.dismiss(animated: true, completion: nil);
            }
        }
    }
    
    @IBAction func onResetPressed(_ sender: Any) {
        //TODO: AlertController
        SeaCatClient.reset();
    }
    
}
